File: sdk-typescript/packages/prism/src/prism.ts
/**
 * WAVE SDK - Prism API
 *
 * Virtual Device Bridge - present network AV sources (NDI, ONVIF, VISCA, Dante)
 * as standard USB UVC/UAC devices to conferencing apps.
 */

import type { WaveClient, PaginationParams, PaginatedResponse, Timestamps } from '@wave-av/core';

// ============================================================================
// Types
// ============================================================================

export type VirtualDeviceType = "camera" | "microphone";
export type DeviceStatus = "created" | "starting" | "running" | "stopping" | "stopped" | "error";
export type SourceProtocol =
  | "ndi"
  | "onvif"
  | "srt"
  | "rtmp"
  | "webrtc"
  | "dante"
  | "cloudflare"
  | "livekit";
export type PTZProtocol = "ndi" | "onvif" | "visca" | "pelcod" | "cgi" | "livekit";

export interface VirtualDevice extends Timestamps {
  id: string;
  organization_id: string;
  name: string;
  type: VirtualDeviceType;
  status: DeviceStatus;
  source_protocol: SourceProtocol;
  source_endpoint: string;
  node_id: string;
  resolution?: { width: number; height: number };
  frame_rate?: number;
  health_score: number;
  ptz_enabled: boolean;
  ptz_protocol?: PTZProtocol;
}

export interface PresetMapping extends Timestamps {
  id: string;
  device_id: string;
  slot_number: number;
  preset_name: string;
  preset_token: string;
  protocol: PTZProtocol;
  transition_speed: number;
}

export interface DeviceHealth {
  device_id: string;
  status: "healthy" | "degraded" | "critical" | "offline";
  latency_ms: number;
  dropped_frames: number;
  fps: number;
  cpu_usage: number;
  source_connected: boolean;
  driver_connected: boolean;
}

export interface DiscoveredSource {
  id: string;
  name: string;
  protocol: string;
  address: string;
  supports_ptz: boolean;
  capabilities: string[];
  discovered_at: string;
}

export interface CreateDeviceRequest {
  name: string;
  type: VirtualDeviceType;
  source_protocol: SourceProtocol;
  source_endpoint: string;
  node_id: string;
  resolution?: { width: number; height: number };
  frame_rate?: number;
  ptz_enabled?: boolean;
  ptz_protocol?: PTZProtocol;
  metadata?: Record<string, unknown>;
}

export interface UpdateDeviceRequest {
  name?: string;
  source_endpoint?: string;
  resolution?: { width: number; height: number };
  frame_rate?: number;
  metadata?: Record<string, unknown>;
}

export interface SetPresetRequest {
  slot_number: number;
  preset_name: string;
  preset_token: string;
  protocol: PTZProtocol;
  transition_speed?: number;
}

export interface ListDevicesParams extends PaginationParams {
  type?: VirtualDeviceType;
  status?: DeviceStatus;
  node_id?: string;
  source_protocol?: SourceProtocol;
  order_by?: string;
  order?: "asc" | "desc";
}

// ============================================================================
// Prism API - Virtual Device Bridge
// ============================================================================

/**
 * Virtual Device Bridge - present network AV sources (NDI, ONVIF, VISCA, Dante)
 * as standard USB UVC/UAC devices to conferencing apps (Zoom, Teams, Meet).
 *
 * @example
 * ```typescript
 * const device = await wave.prism.createDevice({ name: 'PTZ Cam', type: 'camera', source_protocol: 'ndi', source_endpoint: 'NDI-1', node_id: 'node_abc' });
 * await wave.prism.startDevice(device.id);
 * await wave.prism.setPreset(device.id, { slot_number: 1, preset_name: 'Wide', preset_token: 'p1', protocol: 'ndi' });
 * const sources = await wave.prism.discoverSources({ protocols: ['ndi', 'onvif'] });
 * ```
 */
export class PrismAPI {
  private readonly client: WaveClient;
  private readonly basePath = "/v1/prism";

  constructor(client: WaveClient) {
    this.client = client;
  }

  async createDevice(request: CreateDeviceRequest): Promise<VirtualDevice> {
    return this.client.post<VirtualDevice>(`${this.basePath}/devices`, request);
  }

  async getDevice(deviceId: string): Promise<VirtualDevice> {
    return this.client.get<VirtualDevice>(`${this.basePath}/devices/${deviceId}`);
  }

  async updateDevice(deviceId: string, request: UpdateDeviceRequest): Promise<VirtualDevice> {
    return this.client.patch<VirtualDevice>(`${this.basePath}/devices/${deviceId}`, request);
  }

  async removeDevice(deviceId: string): Promise<void> {
    await this.client.delete(`${this.basePath}/devices/${deviceId}`);
  }

  async listDevices(params?: ListDevicesParams): Promise<PaginatedResponse<VirtualDevice>> {
    return this.client.get<PaginatedResponse<VirtualDevice>>(`${this.basePath}/devices`, {
      params: params as Record<string, string | number | boolean | undefined>,
    });
  }

  async startDevice(deviceId: string): Promise<VirtualDevice> {
    return this.client.post<VirtualDevice>(`${this.basePath}/devices/${deviceId}/start`);
  }

  async stopDevice(deviceId: string): Promise<VirtualDevice> {
    return this.client.post<VirtualDevice>(`${this.basePath}/devices/${deviceId}/stop`);
  }

  async getHealth(deviceId: string): Promise<DeviceHealth> {
    return this.client.get<DeviceHealth>(`${this.basePath}/devices/${deviceId}/health`);
  }

  async discoverSources(options?: {
    protocols?: string[];
    subnet?: string;
    timeout?: number;
  }): Promise<DiscoveredSource[]> {
    return this.client.post<DiscoveredSource[]>(`${this.basePath}/discovery`, options);
  }

  async getPresets(deviceId: string): Promise<PresetMapping[]> {
    return this.client.get<PresetMapping[]>(`${this.basePath}/devices/${deviceId}/presets`);
  }

  async setPreset(deviceId: string, request: SetPresetRequest): Promise<PresetMapping> {
    return this.client.put<PresetMapping>(`${this.basePath}/devices/${deviceId}/presets`, request);
  }

  async removePreset(deviceId: string, slotNumber: number): Promise<void> {
    await this.client.delete(`${this.basePath}/devices/${deviceId}/presets/${slotNumber}`);
  }

  async recallPreset(deviceId: string, slotNumber: number): Promise<void> {
    await this.client.post(`${this.basePath}/devices/${deviceId}/presets/${slotNumber}/recall`);
  }
}

export function createPrismAPI(client: WaveClient): PrismAPI {
  return new PrismAPI(client);
}

File: sdk-python/wave/prism.py
"""WAVE SDK - Prism API. Virtual Device Bridge for network AV to USB."""
from __future__ import annotations
from typing import Any
from pydantic import BaseModel
from wave.client import WaveClient

class VirtualDevice(BaseModel):
    id: str; organization_id: str; name: str; type: str; status: str; source_protocol: str; source_endpoint: str; node_id: str; resolution: dict | None = None; frame_rate: int | None = None; health_score: float = 100; ptz_enabled: bool = False; created_at: str; updated_at: str

class PresetMapping(BaseModel):
    id: str; device_id: str; slot_number: int; preset_name: str; preset_token: str; protocol: str; transition_speed: float = 1.0; created_at: str; updated_at: str

class PrismAPI:
    def __init__(self, client: WaveClient): self._client = client; self._base = "/v1/prism"
    def create_device(self, name: str, type: str, source_protocol: str, source_endpoint: str, node_id: str, **kwargs: Any) -> VirtualDevice: return VirtualDevice(**self._client.post(f"{self._base}/devices", json={"name": name, "type": type, "source_protocol": source_protocol, "source_endpoint": source_endpoint, "node_id": node_id, **kwargs}))
    def get_device(self, device_id: str) -> VirtualDevice: return VirtualDevice(**self._client.get(f"{self._base}/devices/{device_id}"))
    def update_device(self, device_id: str, **kwargs: Any) -> VirtualDevice: return VirtualDevice(**self._client.patch(f"{self._base}/devices/{device_id}", json=kwargs))
    def remove_device(self, device_id: str) -> None: self._client.delete(f"{self._base}/devices/{device_id}")
    def list_devices(self, **params: Any) -> dict: return self._client.get(f"{self._base}/devices", params={k: v for k, v in params.items() if v is not None})
    def start_device(self, device_id: str) -> VirtualDevice: return VirtualDevice(**self._client.post(f"{self._base}/devices/{device_id}/start"))
    def stop_device(self, device_id: str) -> VirtualDevice: return VirtualDevice(**self._client.post(f"{self._base}/devices/{device_id}/stop"))
    def get_health(self, device_id: str) -> dict: return self._client.get(f"{self._base}/devices/{device_id}/health")
    def discover_sources(self, **kwargs: Any) -> list[dict]: return self._client.post(f"{self._base}/discovery", json=kwargs if kwargs else None)
    def get_presets(self, device_id: str) -> list[PresetMapping]: return [PresetMapping(**p) for p in self._client.get(f"{self._base}/devices/{device_id}/presets")]
    def set_preset(self, device_id: str, slot_number: int, preset_name: str, preset_token: str, protocol: str, **kwargs: Any) -> PresetMapping: return PresetMapping(**self._client.put(f"{self._base}/devices/{device_id}/presets", json={"slot_number": slot_number, "preset_name": preset_name, "preset_token": preset_token, "protocol": protocol, **kwargs}))
    def remove_preset(self, device_id: str, slot_number: int) -> None: self._client.delete(f"{self._base}/devices/{device_id}/presets/{slot_number}")
    def recall_preset(self, device_id: str, slot_number: int) -> None: self._client.post(f"{self._base}/devices/{device_id}/presets/{slot_number}/recall")
